### PR TITLE
Optionally aggregate metrics for custom step method models in `Trainer.evaluate()`

### PR DESCRIPTION
**Pull request by LarsKue**
Currently, models that use a custom `test_step` function do not support metric aggregation in `evaluate()`. Instead, the method simply returns metrics from the last step:

https://github.com/keras-team/keras/blob/785c9b0b59be9d7921f48562998d4b0257e24ee9/keras/src/trainers/trainer.py#L1013-L1015

This can lead to user confusion, e.g. in https://github.com/bayesflow-org/bayesflow/issues/481.

This PR introduces a very primitive way to aggregate metrics, which is controllable by a boolean flag in the `evaluate` method: `aggregate=False` is set to `False` by default to preserve backward-compatibility. The current idea is to simply sum all metrics and divide by the total number of steps taken, which works when all steps return any PyTree of numerical (tensor or primitive) metrics, which must be consistent across all steps taken.

I understand that this code might still be too simple, lack extensibility to multi-device synchronization, or be inconsistent with other parts of the library. I am open to modifying the PR, given feedback.

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR. Is there an alternative way to solve this problem? Other backends don't have this issue.

**Comment by LarsKue** [comment]
@fchollet Could you elaborate? I think the issue exists independently of the backend.

In terms of options, I believe we could also do something along the lines of `tree_concatenate` and then `tree_mean` or let the user supply an aggregation function. This requires slightly more memory, but would also allow different steps to return different metrics.

**Comment by fchollet** [comment]
Aggregation should be handled by stateful `Metric` instances -- it's not something that you need extra logic for.

The idea being that at each test step, you update the `Metric` instances with the metric value for the last batch, and at the end, you query the `Metric` instances to get the aggregated (average) value.

**Comment by LarsKue** [comment]
@fchollet I agree that this could and should be handled by the metrics themselves. However, the current `keras.Metric` only supports a supervised interface using `y_pred, y_true`. Thus, models that circumvent the mostly supervised-learning oriented interface by implementing a custom `train_step, test_step` will encounter this issue.

In that case, should we fall back to a conversation about extending the support for arbitrary gradient-based learning in keras? I.e. `loss = f(data)` rather than `loss = loss_fn(f(data), label)`.

**Comment by fchollet** [comment]
> However, the current keras.Metric only supports a supervised interface using y_pred, y_true. Thus, models that circumvent the mostly supervised-learning oriented interface by implementing a custom train_step, test_step will encounter this issue.

You can just use a `Mean` metric which will aggregate any tensor (it will even work with non-scalars).

Metric reduction across an epoch (or across one call to `evaluate()`) should be done by metric objects in any case.
